packages/saslprep/src/index.ts

```diff
@@ -0,0 +1,152 @@
+    const before = input.charCodeAt(i);
+
+    if (before >= 0xd800 && before <= 0xdbff && size > i + 1) {
+      const next = input.charCodeAt(i + 1);
+
+      if (next >= 0xdc00 && next <= 0xdfff) {
+        codepoints.push((before - 0xd800) * 0x400 + next - 0xdc00 + 0x10000);
+        i += 1;
+        continue;
+      }
+    }
+
+    codepoints.push(before);
+  }
+
+  return codepoints;
+}
+
+/**
+ * SASLprep.
+ */
+function saslprep(
+  input: string,
+  opts: { allowUnassigned?: boolean } = {}
+): string {
+  if (typeof input !== 'string') {
+    throw new TypeError('Expected string.');
+  }
+
+  if (input.length === 0) {
+    return '';
+  }
+
+  // 1. Map
+  const mapped_input = toCodePoints(input)
+    // 1.1 mapping to space
+    .map((character) => (mapping2space.get(character) ? 0x20 : character))
+    // 1.2 mapping to nothing
+    .filter((character) => !mapping2nothing.get(character));
+
+  // 2. Normalize
+  const normalized_input = String.fromCodePoint
+    .apply(null, mapped_input)
+    .normalize('NFKC');
+
+  const normalized_map = toCodePoints(normalized_input);
+
+  // 3. Prohibit
+  const hasProhibited = normalized_map.some((character) =>
+    prohibited_characters.get(character)
+  );
+
+  if (hasProhibited) {
+    throw new Error(
+      'Prohibited character, see https://tools.ietf.org/html/rfc4013#section-2.3'
+    );
+  }
+
+  // Unassigned Code Points
+  if (opts.allowUnassigned !== true) {
+    const hasUnassigned = normalized_map.some((character) =>
+      unassigned_code_points.get(character)
+    );
+
+    if (hasUnassigned) {
+      throw new Error(
+        'Unassigned code point, see https://tools.ietf.org/html/rfc4013#section-2.5'
+      );
+    }
+  }
+
+  // 4. check bidi
+
+  const hasBidiRAL = normalized_map.some((character) =>
+    bidirectional_r_al.get(character)
+  );
+
+  const hasBidiL = normalized_map.some((character) =>
+    bidirectional_l.get(character)
+  );
+
+  // 4.1 If a string contains any RandALCat character, the string MUST NOT
+  // contain any LCat character.
+  if (hasBidiRAL && hasBidiL) {
+    throw new Error(
+      'String must not contain RandALCat and LCat at the same time,' +
+        ' see https://tools.ietf.org/html/rfc3454#section-6'
+    );
+  }
+
+  /**
+   * 4.2 If a string contains any RandALCat character, a RandALCat
+   * character MUST be the first character of the string, and a
+   * RandALCat character MUST be the last character of the string.
+   */
+
+  const isFirstBidiRAL = bidirectional_r_al.get(
+    getCodePoint(first(normalized_input))!
+  );
+  const isLastBidiRAL = bidirectional_r_al.get(
+    getCodePoint(last(normalized_input))!
+  );
+
+  if (hasBidiRAL && !(isFirstBidiRAL && isLastBidiRAL)) {
+    throw new Error(
+      'Bidirectional RandALCat character must be the first and the last' +
+        ' character of the string, see https://tools.ietf.org/html/rfc3454#section-6'
+    );
+  }
+
+  return normalized_input;
+}
+
+export = saslprep;
```

Comment:
Could this be a named or default export now? or is that too much friction?

Reply:
We could do that, or we could add something like

```suggestion
saslprep.saslprep = saslprep;
saslprep.default = saslprep;
export = saslprep;
```

for compatibility. No strong feelings here, I mostly just wanted to keep everything as-is in terms of code and behavior for this PR.

Reply:
Fine with either approach, I know for us at least migrating to a named export will be easy even if we want to backport the migration to this dependency to previous driver versions. Changing this makes it easier to not use tslib/helpers which we are still avoiding (I can't speak for the future)

Reply:
Ok, then I’ve made this change here as suggested above, I think that should be fine as an addition to the initial feature :+1: